client/packages/purchase_orders/src/api/hooks/usePurchaseOrder.ts

```diff
@@ -0,0 +1,85 @@
+import {
+  FnUtils,
+  InsertPurchaseOrderInput,
+  useMutation,
+  useQuery,
+} from '@openmsupply-client/common';
+import { usePurchaseOrderGraphQL } from '../usePurchaseOrderGraphQL';
+import { LIST, PURCHASE_ORDER } from './keys';
+import { PurchaseOrderFragment } from '../operations.generated';
+
+export const usePurchaseOrder = (purchaseOrderId?: string) => {
+  const { purchaseOrderApi, storeId } = usePurchaseOrderGraphQL();
+
+  const queryKey = [LIST, PURCHASE_ORDER, storeId, purchaseOrderId];
+
+  // QUERY
+
+  const queryFn = async (): Promise<PurchaseOrderFragment | void> => {
+    if (!purchaseOrderId) return;
+
+    const result = await purchaseOrderApi.purchaseOrderById({
+      purchaseOrderId,
+      storeId,
+    });
+    const purchaseOrder = result?.purchaseOrder;
+    if (purchaseOrder.__typename === 'PurchaseOrderNode') return purchaseOrder;
+    else {
+      console.error('No purchase order found', purchaseOrderId);
+      throw new Error(`Could not find purchase order ${purchaseOrderId}`);
+    }
+  };
+
+  const { data, isLoading, isError } = useQuery({
+    queryKey,
+    queryFn,
+    enabled: !!purchaseOrderId,
+  });
+
+  // UPDATE
+
+  // CREATE
+
+  const {
+    mutateAsync: createMutation,
+    isLoading: isCreating,
+    error: createError,
+  } = useCreate();
+
+  const create = async (supplierId: string) => {
+    const id = FnUtils.generateUUID();
+
+    const result = await createMutation({ id, supplierId });
+    return result;
+  };
+
+  return {
+    query: { data, isLoading, isError },
+    create: { create, isCreating, createError },
+  };
+};
+
+const useCreate = () => {
+  const { purchaseOrderApi, storeId, queryClient } = usePurchaseOrderGraphQL();
+
+  const mutationFn = async (
+    input: InsertPurchaseOrderInput
+  ): Promise<string> => {
+    const result =
+      (await purchaseOrderApi.insertPurchaseOrder({
+        input,
+        storeId,
+      })) || {};
+
+    const { insertPurchaseOrder } = result;
+
+    if (insertPurchaseOrder.id) return insertPurchaseOrder.id;
+
+    throw new Error('Could not insert purchase order');
+  };
```

Comment:
This is cool - I was often having difficulty getting try catch blocks working, but you can just throw here and catch elsewhere!

Reply:
Yeah, we don't really have a consistent pattern around where to `throw` in the front-end, but I generally find putting the `catch` at the level you're actually going to deal with it (e.g. display a notification?) is nicest.